internal/pfcpsim/state.go

```diff
@@ -0,0 +1,116 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2022 Open Networking Foundation
+ */
+
+package pfcpsim
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/omec-project/pfcpsim/pkg/pfcpsim"
+)
+
+var (
+	activeSessions     = make(map[int]*pfcpsim.PFCPSession, 0)
+	lockActiveSessions = new(sync.Mutex)
+
+	remotePeerAddress string
+	upfN3Address      string
+
+	// Emulates 5G SMF/ 4G SGW
+	sim                 *pfcpsim.PFCPClient
+	remotePeerConnected bool
+)
+
+func insertSession(index int, session *pfcpsim.PFCPSession) {
+	lockActiveSessions.Lock()
+	defer lockActiveSessions.Unlock()
+
+	activeSessions[index] = session
+}
+
+func getSession(index int) (*pfcpsim.PFCPSession, bool) {
+	element, ok := activeSessions[index]
+	return element, ok
+}
+
+func deleteSession(index int) {
+	lockActiveSessions.Lock()
+	defer lockActiveSessions.Unlock()
+
+	delete(activeSessions, index)
+}
+
+func isConfigured() bool {
+	if upfN3Address != "" && remotePeerAddress != "" {
+		return true
+	}
+
+	return false
+}
+
+func isRemotePeerConnected() bool {
+	return remotePeerConnected
+}
+
+func connectPFCPSim() error {
+	if sim == nil {
+		localAddress, err := getLocalAddress()
+		if err != nil {
+			return err
+		}
+
+		sim = pfcpsim.NewPFCPClient(localAddress)
+	}
+
+	err := sim.ConnectN4(remotePeerAddress)
+	if err != nil {
+		return err
+	}
+
+	remotePeerConnected = true
+
+	return nil
+}
+
+// getLocalAddress retrieves local address in string format to use when establishing a connection with PFCP agent
+func getLocalAddress() (string, error) {
```

Comment:
Isn't there a better way to get local address than using CLI? See: https://github.com/omec-project/upf-epc/blob/master/pfcpiface/utils.go#L119

Reply:
This script automatically retrieves the internet-connected interface basically not caring about entering an interface name.

with your proposal the user must enter the interface name using the `Configure` RPC passing the server's local interface.
If the pfcpsimctl lives in the same container it's not a big deal, but, since this approach allows you to connect from wherever you want (e.g. using pfcpsimctl in your local machine giving commands to gRPC Server living in a VM as I'm using it nowadays), in general you don't know the server's internet interface. Sure we could pass it as a flag when starting the server, but the purpose of `Configure` was to avoid starting the server with flags and let the user configure some parameters and be able to override them at runtime. 

Having said that, if you think it's better to pass the interface name I can change it right away 😄